File: Lintcode 016 - Permutations II.java
Lintcode 016 - Permutations II.java

class Solution {
    /**
     * @param nums: A list of integers.
     * @return: A list of unique permutations.
     */
    public ArrayList<ArrayList<Integer>> permuteUnique(ArrayList<Integer> nums) {
        // write your code here
        ArrayList<ArrayList<Integer>> result = new ArrayList<ArrayList<Integer>>();
        if(nums == null || nums.size() == 0) {
            return result;
        }
        
        ArrayList<Integer> list = new ArrayList<Integer>();
        boolean[] used = new boolean[nums.length];
        
        Collections.sort(nums);
        dfs(result, list, nums, used);
        
        return result;
    }
    
    public void dfs(ArrayList<ArrayList<Integer>> result, ArrayList<Integer> list, ArrayList<Integer> nums, boolean[] used) {
        
        if(list.size() == nums.size()) {
            result.add(new ArrayList(list));
            return; //go back to 上一步
        }
        
        for(int i = 0; i < nums.size(); i++) {
            //连续3个 && 条件，一个都不能少
            if(i > 0 && !used[i-1] && nums.get(i) == nums.get(i-1)) {
                continue;
            }

            if(used[i]) {
                continue;
            }
            
            list.add(nums.get(i));
            used[i] = true;
            dfs(result, list, nums, used);
            list.remove(list.size()-1);
            used[i] = false;

        }
    }
}

File: Lintcode 016 - Permutations II.cs
// Lintcode 016 - Permutations II
using System;
using System.Collections;
using System.Collections.Generic;

namespace CSharp
{
    class Program{
        static void Main()
        {
            int[] nums = new int[4]{1,1,3,4};
            List<List<int>> result = new List<List<int>>{};   
            List<int> list = new List<int>{};
            bool[] used = new bool[4];

            Permutation2(nums, result, list, used);

            foreach(var clist in result)
            {
                foreach(var num in clist)
                {
                    Console.Write(num + ", ");
                }
                Console.WriteLine(" ");
            }
        }

        static void Permutation2(int[] nums, List<List<int>> result, List<int> list, bool[] used)
        {

            if(list.Count == nums.Length)
            {
                result.Add(new List<int>(list));
                return;
            }

            for(int i = 0; i< nums.Length; i++)
            {
                if(i > 0 && nums[i - 1] == nums[i] && !used[i - 1])
                {
                    continue;
                }

                if(used[i])
                {
                    continue;
                }

                used[i] = true;
                list.Add(nums[i]);
                Permutation2(nums, result, list, used);
                list.Remove(list[list.Count - 1]);
                used[i] = false;
            }
        }
    }
}
